migration/1670429143091-createGivPowerHistoricTablesProcedure.ts

```diff
@@ -0,0 +1,73 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class createGivPowerHistoricTablesProcedure1670429143091
+  implements MigrationInterface
+{
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    // power boosting
+    await queryRunner.query(`
+            CREATE OR REPLACE PROCEDURE PUBLIC."TAKE_POWER_BOOSTING_SNAPSHOT_HISTORY"() LANGUAGE 'sql' AS $BODY$
+                WITH snapshot_entity AS (
+                    DELETE FROM "power_boosting_snapshot" AS pbs
+                    WHERE pbs."powerSnapshotId" IN (
+                        SELECT "snapshot"."id"
+                        FROM "power_snapshot" AS "snapshot", "power_round" AS "powerRound"
+                        WHERE "snapshot"."roundNumber" < "powerRound"."round" - 3
+                    )
+                    RETURNING pbs."id", pbs."userId", pbs."projectId", pbs."powerSnapshotId", pbs."percentage"
+                )
+                INSERT INTO "power_boosting_snapshot_history"
+                SELECT se."id", se."userId", se."projectId", se."powerSnapshotId", se."percentage"
+                FROM snapshot_entity AS se
+            $BODY$;
+        `);
+
+    // power balance
+    await queryRunner.query(`
+            CREATE OR REPLACE PROCEDURE PUBLIC."TAKE_POWER_BALANCE_SNAPSHOT_HISTORY"() LANGUAGE 'sql' AS $BODY$
+                WITH snapshot_entity AS (
+                    DELETE FROM "power_balance_snapshot" AS pbs
+                    WHERE pbs."powerSnapshotId" IN (
+                        SELECT "snapshot"."id"
+                        FROM "power_snapshot" AS "snapshot", "power_round" AS "powerRound"
+                        WHERE "snapshot"."roundNumber" < "powerRound"."round" - 3
+                    )
+                    RETURNING pbs."id", pbs."userId", pbs."balance", pbs."powerSnapshotId"
+                )
+                INSERT INTO "power_balance_snapshot_history"
+                SELECT se."id", se."userId", se."balance", se."powerSnapshotId"
+                FROM snapshot_entity AS se
+            $BODY$;
+        `);
+
+    // power snapshot
+    await queryRunner.query(`
+            CREATE OR REPLACE PROCEDURE PUBLIC."TAKE_POWER_SNAPSHOT_HISTORY"() LANGUAGE 'sql' AS $BODY$
+                WITH snapshot_entity AS (
+                    DELETE FROM "power_snapshot" AS ps
+                    WHERE ps."id" IN (
+                        SELECT "snapshot"."id"
+                        FROM "power_snapshot" AS "snapshot", "power_round" AS "powerRound"
+                        WHERE "snapshot"."roundNumber" < "powerRound"."round" - 3
+                    )
+                    RETURNING ps."id", ps."time", ps."blockNumber", ps."roundNumber", ps."synced"
+                )
+                INSERT INTO "power_snapshot_history"
+                SELECT se."id", se."time", se."blockNumber", se."roundNumber", se."synced"
+                FROM snapshot_entity AS se
+            $BODY$;
+        `);
```

Comment:
I think it would be better to aggregate all these actions in a single procedure. I mean instead of three procedures, only define one which does the whole. We can do these three as a transaction also to be sure all is done or nothing will happen
https://www.postgresql.org/docs/current/tutorial-transactions.html

The main reason for this is that we will have to define separate pg_cron job and clutter it's table.

Reply:
Updated them into one! 
The only thing I couldn't do is put a BEGIN END keywords.
It kept saying the logs that You cannot put Begin inside a procedure.

Reply:
```
BEGIN is not allowed in an SQL function
```